odb_extractor_atoms.py

```diff
@@ -59,98 +223,297 @@ def select_obsclass(classes):
 # --------------
 
 
-def findatoms(observation):
-    '''Analyze a json observing sequence from the ODB and define atoms.'''
+def autocorr_lag(x, plot=False):
+    '''Test for patterns with auto-correlation'''
+
+    # Auto correlation
+    result = np.correlate(x, x, mode='full')
+    corrmax = np.max(result)
+    if corrmax != 0.0:
+        result = result / corrmax
+    #     print(result)
+    #     print(result.size, result.size//2)
+    if plot:
+        plt.plot(result[result.size // 2:])
+
+    # Pattern offset using first prominent peak
+    # lag = np.argmax(result[result.size//2 + 1:]) + 1
+    peaks, prop = find_peaks(result[result.size // 2:], height=(0, None), prominence=(0.25, None))
+    #     print(peaks)
+    #     print(prop)
+    lag = 0
+    if len(peaks) > 0:
+        lag = peaks[0]
+
+    return lag
+
+# --------------
+
+
+def findatoms(observation, verbose=False, ws=None, fid=sys.stdout):
+    """Analyze a json observing sequence from the ODB and define atoms."""
 
     classes = []
+    guiding = []
     qastates = []
     atoms = []
     natom = 0
-    nabba = 0
 
     # Make dictionary out of obslog to get QA state
     obslog = {}
-    for log_entry in observation['obsLog']:
-        obslog[log_entry['label']] = {'qaState': log_entry['qaState'], 'filename': log_entry['filename']}
-    datalabels = list(obslog.keys())
+    datalabels = []
+    if 'obsLog' in observation.keys():
+        for log_entry in observation['obsLog']:
+            obslog[log_entry['label']] = {'qaState': log_entry['qaState'], 'filename': log_entry['filename']}
+        datalabels = list(obslog.keys())
 
     # Sequence analysis
     sequence = observation['sequence']
     nsteps = len(sequence)
 
+    # First pass to analyze offsets and exptimes/coadds
+    do_not_split = False
```

Comment:
We don't need this here: we should move it down to where it is used. See my suggestion below.